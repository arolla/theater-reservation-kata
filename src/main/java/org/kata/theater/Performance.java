package org.kata.theater;

import java.time.LocalDateTime;

public class Performance {

    long id;
    String play; // "The CICD - Corneille", "Les fourberies de Scala - Moli√®re"

    LocalDateTime startTime;

    LocalDateTime endTime;

    String performanceNature; // can be "PREVIEW", "PREMIERE", etc
}
