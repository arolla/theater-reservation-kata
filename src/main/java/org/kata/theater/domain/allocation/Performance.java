package org.kata.theater.domain.allocation;

import lombok.Value;
import org.kata.theater.domain.reservation.TheaterSession;

@Value
public class Performance {

    long id;

    TheaterSession theaterSession;

    PerformanceNature nature;
}
