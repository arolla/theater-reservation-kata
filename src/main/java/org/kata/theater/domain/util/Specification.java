package org.kata.theater.domain.util;

public interface Specification<T> {

    boolean isSatisfiedBy(T t);
}
