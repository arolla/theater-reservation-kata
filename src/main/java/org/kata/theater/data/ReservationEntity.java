package org.kata.theater.data;

import java.util.Arrays;

public class ReservationEntity {

    private Long reservationId;

    private Long performanceId;

    private String[] seats;

    public String getStatus() {
        return getSeats().length > 0 ? "PENDING" : "ABORTED";
    }

    public Long getReservationId() {
        return reservationId;
    }

    public void setReservationId(Long reservationId) {
        this.reservationId = reservationId;
    }

    public Long getPerformanceId() {
        return performanceId;
    }

    public void setPerformanceId(Long performanceId) {
        this.performanceId = performanceId;
    }

    public String[] getSeats() {
        return seats;
    }

    public void setSeats(String[] seats) {
        this.seats = seats;
    }

    @Override
    public String toString() {
        return "Reservation{" +
                "reservationId=" + reservationId +
                ", status='" + getStatus() + '\'' +
                ", seats=" + Arrays.toString(seats) +
                '}';
    }
}
