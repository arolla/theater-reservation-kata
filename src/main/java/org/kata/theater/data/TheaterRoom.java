package org.kata.theater.data;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class TheaterRoom {

    private Zone[] zones;

    public TheaterRoom(Zone[] zones) {
        this.zones = zones;
    }

    public Zone[] getZones() {
        return zones;
    }

    public void setZones(Zone[] zones) {
        this.zones = zones;
    }

    public List<Seat> allSeats() {
        return Arrays.stream(zones)
                .map(Zone::seats)
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }

    public List<String> freeSeats() {
        return allSeats().stream()
                .filter(Seat::isFree)
                .map(Seat::getSeatId)
                .collect(Collectors.toList());
    }
}
