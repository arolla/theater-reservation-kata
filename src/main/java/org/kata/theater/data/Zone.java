package org.kata.theater.data;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class Zone {
    private Row[] rows;

    private String category; // STANDARD, PREMIUM

    public Zone(Row[] rows, String category) {
        this.rows = rows;
        this.category = category;
    }

    public Row[] getRows() {
        return rows;
    }

    public void setRows(Row[] rows) {
        this.rows = rows;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public List<Seat> seats() {
        return Arrays.stream(rows)
                .map(row -> Arrays.asList(row.getSeats()))
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }
}
