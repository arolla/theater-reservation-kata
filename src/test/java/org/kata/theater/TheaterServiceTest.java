package org.kata.theater;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.kata.theater.data.Performance;

import java.time.LocalDate;
import java.time.Month;
import java.util.List;

class TheaterServiceTest {

    private TheaterService theaterService;

    @BeforeEach
    void setUp() {
        theaterService = new TheaterService();
    }

    @Test
    void reserve_once_on_premiere_performance() {
        Performance performance = new Performance();
        performance.id = 1L;
        performance.play = "The CICD by Corneille";
        performance.startTime = LocalDate.of(2023, Month.APRIL, 22).atTime(21, 0);
        performance.performanceNature = "PREMIERE";
        String reservation = theaterService.reservation(1L, 4, "STANDARD",
                performance);
        Approvals.verify(reservation);
    }

    @Test
    void reserve_once_on_premiere_performance_with_premium_category() {
        Performance performance = new Performance();
        performance.id = 1L;
        performance.play = "The CICD by Corneille";
        performance.startTime = LocalDate.of(2023, Month.APRIL, 22).atTime(21, 0);
        performance.performanceNature = "PREMIERE";
        String reservation = theaterService.reservation(1L, 4, "PREMIUM",
                performance);
        Approvals.verify(reservation);
    }

    @Test
    void cancel_then_reserve_on_premiere_performance_with_standard_category() {
        theaterService.cancelReservation("45", 1L, List.of("B2"));
        Performance performance = new Performance();
        performance.id = 1L;
        performance.play = "The CICD by Corneille";
        performance.startTime = LocalDate.of(2023, Month.APRIL, 22).atTime(21, 0);
        performance.performanceNature = "PREMIERE";
        String reservation = theaterService.reservation(1L, 4, "STANDARD",
                performance);
        Approvals.verify(reservation);
    }

    @Test
    void reserve_twice_on_premiere_performance() {
        Performance performance = new Performance();
        performance.id = 1L;
        performance.play = "The CICD by Corneille";
        performance.startTime = LocalDate.of(2023, Month.APRIL, 22).atTime(21, 0);
        performance.performanceNature = "PREMIERE";
        String reservation1 = theaterService.reservation(1L, 4, "STANDARD",
                performance);
        String reservation2 = theaterService.reservation(1L, 5, "STANDARD",
                performance);
        Approvals.verify(reservation2);
    }

    @Test
    void reservation_failed_on_preview_performance() {
        Performance performance = new Performance();
        performance.id = 2L;
        performance.play = "Les fourberies de Scala - Molière";
        performance.startTime = LocalDate.of(2023, Month.MARCH, 21).atTime(21, 0);
        performance.performanceNature = "PREVIEW";

        String reservation = theaterService.reservation(2L, 4, "STANDARD",
                performance);
        Approvals.verify(reservation);
    }

    @Test
    void reservation_failed_on_premiere_performance() {
        Performance performance = new Performance();
        performance.id = 3L;
        performance.play = "DOM JSON - Molière";
        performance.startTime = LocalDate.of(2023, Month.MARCH, 21).atTime(21, 0);
        performance.performanceNature = "PREMIERE";

        String reservation = theaterService.reservation(2L, 4, "STANDARD",
                performance);
        Approvals.verify(reservation);
    }

}