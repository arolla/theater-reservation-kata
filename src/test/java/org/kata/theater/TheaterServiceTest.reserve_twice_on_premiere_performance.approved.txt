<reservation>
	<performance>
		<play>The CICD by Corneille</play>
		<date>2023-04-22</date>
		<time>21:00</time>
	</performance>
	<reservationId>123456</reservationId>
	<reservationStatus>FULFILLABLE</reservationStatus>
		<seats>
			<seat>C1</seat>
			<seat>C2</seat>
			<seat>C3</seat>
			<seat>C4</seat>
			<seat>C5</seat>
		</seats>
	<seatCategory>STANDARD</seatCategory>
	<totalAmountDue>144.37â‚¬</totalAmountDue>
</reservation>
